src/ast/sym.cc

```diff
@@ -247,27 +296,392 @@ namespace sym
+            // zero argument invoke
+            auto args = ast::node(ast, "args");
+            ast::push_back(ast, args);
+          }
+          break;
+        }
+
+        case "tuple"_:
+        {
+          // (invoke obj typeargs args)
+          auto obj = ast;
+          auto invoke = ast::node(next, "invoke");
+          ast::replace(ast, invoke);
+          ast::push_back(ast, obj);
+          auto typeargs = ast::node(ast, "typeargs");
+          ast::push_back(ast, typeargs);
+          ast::remove(next);
+          ast::rename(next, "args");
+          ast::push_back(ast, next);
+          break;
+        }
+
+        default:
+          return;
+      }
+    }
+  }
+
+  void prec_prefix(ast::Ast& ast, err::Errors& err)
+  {
+    // (call function typeargs obj (args))
+    if (ast && (ast->tag == "op"_))
+    {
+      auto op = ast;
+      auto call = ast::node(ast, "call");
+      ast::replace(ast, call);
+      ast::rename(op, "function");
+      ast::push_back(ast, op);
+      auto next = ast::get_next_in_expr(ast);
+      add_typeargs(ast, next);
+      prec_prefix(next, err);
+
+      if (!next)
+      {
+        err << ast << "Expected an argument to this prefix function."
+            << err::end;
+        return;
+      }
+
+      ast::remove(next);
+      ast::push_back(ast, next);
+      auto args = ast::node(ast, "args");
+      ast::push_back(ast, args);
+    }
+    else
+    {
+      prec_member(ast, err);
+    }
+  }
+
+  void prec_infix(ast::Ast& ast, err::Errors& err)
+  {
+    // (call function typeargs lhs (args rhs))
+    // Infix operators are left associative.
+    prec_prefix(ast, err);
+    auto next = ast::get_next_in_expr(ast);
+    std::string op;
+
+    while (next)
+    {
+      if (next->tag == "assign"_)
+        return;
+
+      if (next->tag != "op"_)
+      {
+        err << next << "Expected an infix operator here." << err::end;
+        return;
+      }
+
+      if (op.empty())
+      {
+        op = next->token;
+      }
+      else if (op != next->token)
+      {
+        err << next << "Use parentheses to establish precedence between " << op
+            << " and " << next->token << " infix operators." << err::end;
+      }
+
+      auto lhs = ast;
+      auto call = ast::node(ast, "call");
+      ast::replace(ast, call);
+      ast::remove(next);
+      ast::rename(next, "function");
+      ast::push_back(ast, next);
+      next = ast::get_next_in_expr(ast);
+      add_typeargs(ast, next);
+      ast::push_back(ast, lhs);
+      auto args = ast::node(ast, "args");
+      ast::push_back(ast, args);
+
+      if (!next)
+      {
+        err << ast << "Expected an expression after this infix operator."
+            << err::end;
+        return;
+      }
+
+      prec_prefix(next, err);
+
+      if (next)
+      {
+        ast::remove(next);
+        ast::push_back(args, next);
+        next = ast::get_next_in_expr(ast);
+      }
+    }
+  }
+
+  void prec_assign(ast::Ast& ast, err::Errors& err)
```

Comment:
I wonder if this might be better named
```C++
prec_for_assign
```
I was thinking it is parsing assignments to start with.  And as it doesn't always require an assign, this was a little confusing. Adding `for` makes it less about there being an assign in my mind.

Reply:
I'm sorry, that doesn't make sense to me. It's the assign rule in the precedence pass. Not sure I want to rename every rewrite rule as `for_$rule` ? There may be no infix in `prec_infix` and no prefix in `prec_prefix`.